### PR TITLE
[AWS SQS] Support strict expiration of SQS batches when writing

### LINKED ISSUE #27483: [Feature Request]: Support strict batch timeouts for AWS SQS batch writes
### What would you like to happen?

Timeouts of SQS batches are checked on arrival of new messages.
In cases where this causes too much delay (short timeouts, irregular/low frequency of messages), strict enforcement of batch timeouts should be done using a separate scheduler thread.

Related to #21429

### Issue Priority

Priority: 3 (nice-to-have improvement)

### Issue Components

- [ ] Component: Python SDK
- [ ] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [X] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### PR DESCRIPTION
**Pull request by mosche**
Support strict expiration of AWS SQS batches when writing.

Closes #27483

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by mosche** [comment]
R: @psolomin 
R: @aromanenko-dev

**Review by psolomin — COMMENTED** [review]
Hello! Sorry for the delayed reply.

I will have time to carefully look into the code within the next 3 days. For now I've ran some basic "smoke" checks, and got a question (see in the comment)

File: sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/sqs/SqsIOWriteBatchesTest.java [hunk 1]
```diff
@@ -187,24 +237,53 @@ public void testWriteBatchesToDynamic() {
...
-    verify(sqs).sendMessageBatch(request("queue", entries[2], entries[3]));
-    verify(sqs).sendMessageBatch(request("queue", entries[4]));
+    // due to added delay, dynamic batches are timed out on arrival of every 2nd msg (per batch)
+    verify(sqs).sendMessageBatch(request("even", entries[0], entries[2]));
+    verify(sqs).sendMessageBatch(request("uneven", entries[1], entries[3]));
+    verify(sqs).sendMessageBatch(request("even", entries[4]));
+  }
+
+  @Test
+  public void testWriteBatchesToDynamicWithStrictTimeout() {
```
[the remark this answers is not in the document]
Comment (mosche) [reply sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/sqs/SqsIOWriteBatchesTest.java:261]: Thanks for having a look @psolomin :)
The intention of this test was just to check no messages are lost / delivered repeatedly if writing a lot but your concern is absolutely right, this doesn't check it's actually using the strict timeout. Will think about it, but likely I'll just add another test that also verifies the timing (with just a few message as done above) 👍

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/sqs/SqsIO.java [hunk 2]
```diff
@@ -612,57 +655,95 @@ private void writeEntries(Batch batch, boolean throwPendingFailures) {
       }
 
       public void finishBundle() throws Throwable {
-        batches.writeAll();
+        if (expirationCheck != null) {
+          expirationCheck.cancel(false);
+          while (true) {
+            try {
+              expirationCheck.get(3, TimeUnit.SECONDS);
```
Comment (aromanenko-dev) [inline sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/sqs/SqsIO.java:662]: nit: extract "3" a constant

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/sqs/SqsIO.java [hunk 3]
```diff
@@ -709,8 +789,9 @@ private class Dynamic extends Batches {
             (queue, batch) -> batch != null && batch.lock(true) ? batch : createLocked(queue);
 
         private final Map<@NonNull String, Batch> batches = new HashMap<>();
+        private final AtomicBoolean submitExpiredRunning = new AtomicBoolean(false);
```
Comment (aromanenko-dev) [inline sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/sqs/SqsIO.java:792]: Why do we need atomic types here if it's not thread safe?

**Review by aromanenko-dev — APPROVED** [review]
LGTM, just a couple of notes

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/sqs/SqsIO.java (hunk 3 above)
Comment (mosche) [reply sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/sqs/SqsIO.java:792]: The class is not entirely thread-safe, but certain operations have to be, e.g. `submitExpired(boolean throwFailures)`

**Pull request merged**